File: frontend/src/lib/logging/factory.ts
/**
 * Logger factory implementation that manages provider selection and logger creation.
 */

import { Logger, LoggerFactory, LogProvider, LogLevel, LogContext, LogEntry } from './interfaces';

export class DefaultLogger implements Logger {
    private component: string;
    private provider: LogProvider;
    private defaultContext: LogContext;

    constructor(component: string, provider: LogProvider, defaultContext: LogContext = {}) {
        this.component = component;
        this.provider = provider;
        this.defaultContext = defaultContext;
    }

    withContext(context: LogContext): Logger {
        const mergedContext = { ...this.defaultContext, ...context };
        return new DefaultLogger(this.component, this.provider, mergedContext);
    }

    private createLogEntry(
        level: LogLevel,
        message: string,
        error?: Error,
        context: LogContext = {}
    ): LogEntry {
        // Merge all context
        const mergedContext: LogContext = {
            ...this.defaultContext,
            ...context,
            component: this.component,
            environment: process.env.NODE_ENV,
            service_name: process.env.NEXT_PUBLIC_SERVICE_NAME,
            service_version: process.env.NEXT_PUBLIC_SERVICE_VERSION,
            user_agent: typeof window !== 'undefined' ? window.navigator.userAgent : undefined,
            url: typeof window !== 'undefined' ? window.location.href : undefined,
            referrer: typeof window !== 'undefined' ? document.referrer : undefined,
        };

        // Get stack trace for source location
        const stack = new Error().stack;
        let sourceFile: string | undefined;
        let sourceLine: number | undefined;
        let sourceFunction: string | undefined;

        if (stack) {
            // Parse stack trace to get caller information
            const lines = stack.split('\n');
            // Skip the first few lines (Error constructor, this method, and the calling log method)
            const callerLine = lines[3] || lines[2] || lines[1];
            if (callerLine) {
                const match = callerLine.match(/at (.+?) \((.+?):(\d+):\d+\)/) ||
                             callerLine.match(/at (.+?):(\d+):\d+/) ||
                             callerLine.match(/(.+?)@(.+?):(\d+):\d+/);
                if (match) {
                    sourceFunction = match[1]?.trim();
                    sourceFile = match[2]?.trim();
                    sourceLine = parseInt(match[3], 10);
                }
            }
        }

        const entry: LogEntry = {
            level,
            message,
            timestamp: new Date(),
            context: mergedContext,
            source_file: sourceFile,
            source_line: sourceLine,
            source_function: sourceFunction
        };

        if (error) {
            entry.error = {
                name: error.name,
                message: error.message,
                stack: error.stack
            };
        }

        return entry;
    }

    debug(message: string, context?: LogContext): void {
        const entry = this.createLogEntry(LogLevel.DEBUG, message, undefined, context);
        this.provider.log(entry).catch(console.error);
    }

    info(message: string, context?: LogContext): void {
        const entry = this.createLogEntry(LogLevel.INFO, message, undefined, context);
        this.provider.log(entry).catch(console.error);
    }

    warn(message: string, context?: LogContext): void {
        const entry = this.createLogEntry(LogLevel.WARNING, message, undefined, context);
        this.provider.log(entry).catch(console.error);
    }

    error(message: string, error?: Error, context?: LogContext): void {
        const entry = this.createLogEntry(LogLevel.ERROR, message, error, context);
        this.provider.log(entry).catch(console.error);
    }

    critical(message: string, error?: Error, context?: LogContext): void {
        const entry = this.createLogEntry(LogLevel.CRITICAL, message, error, context);
        this.provider.log(entry).catch(console.error);
    }

    logRequest(
        method: string,
        url: string,
        status?: number,
        latencyMs?: number,
        responseSize?: number,
        context?: LogContext
    ): void {
        const level = status && status >= 500 ? LogLevel.ERROR :
                     status && status >= 400 ? LogLevel.WARNING :
                     LogLevel.INFO;

        let message = `${method} ${url}`;
        if (status) {
            message += ` -> ${status}`;
        }

        const entry = this.createLogEntry(level, message, undefined, context);
        entry.http_method = method;
        entry.http_url = url;
        entry.http_status = status;
        entry.http_latency_ms = latencyMs;
        entry.http_response_size = responseSize;

        this.provider.log(entry).catch(console.error);
    }
}

export class DefaultLoggerFactory implements LoggerFactory {
    private provider: LogProvider;

    constructor(provider: LogProvider) {
        this.provider = provider;
    }

    createLogger(component: string, defaultContext?: LogContext): Logger {
        return new DefaultLogger(component, this.provider, defaultContext || {});
    }

    setProvider(provider: LogProvider): void {
        if (this.provider) {
            this.provider.close().catch(console.error);
        }
        this.provider = provider;
    }

    getProvider(): LogProvider {
        return this.provider;
    }
}

// Global factory instance
let _loggerFactory: LoggerFactory | null = null;

export function createLoggerFactory(provider: LogProvider): LoggerFactory {
    const factory = new DefaultLoggerFactory(provider);
    _loggerFactory = factory;
    return factory;
}

export function getLogger(component: string, defaultContext?: LogContext): Logger {
    if (!_loggerFactory) {
        throw new Error('Logger factory not initialized. Call createLoggerFactory() first.');
    }
    
    return _loggerFactory.createLogger(component, defaultContext);
}

export function setProvider(provider: LogProvider): void {
    if (!_loggerFactory) {
        throw new Error('Logger factory not initialized. Call createLoggerFactory() first.');
    }
    
    _loggerFactory.setProvider(provider);
}

File: shared/python/glogger/factory.py
"""
Logger factory implementation that manages provider selection and logger creation.
"""

import os
import sys
import traceback
from datetime import datetime
from typing import Any, Dict

from .interfaces import (
    LogContext,
    LogEntry,
    Logger,
    LoggerFactory,
    LogLevel,
    LogProvider,
)


class DefaultLogger(Logger):
    """
    Default logger implementation that works with any LogProvider.

    This logger provides the application-facing API and delegates actual
    logging to the configured provider.
    """

    def __init__(self, component: str, provider: LogProvider, default_context: Dict[str, Any]):
        self.component = component
        self.provider = provider
        self.default_context = default_context

    def with_context(self, **context_fields) -> Logger:
        """Create a new logger with additional context."""
        merged_context = {**self.default_context, **context_fields}
        return DefaultLogger(self.component, self.provider, merged_context)

    def _create_log_entry(
        self, level: LogLevel, message: str, exception: Exception | None = None, **context
    ) -> LogEntry:
        """Create a log entry with source location and context."""
        # Get caller information
        frame = sys._getframe(2)  # Skip this method and the calling log method

        # Merge all context
        merged_context = {**self.default_context, **context}

        log_context = LogContext(
            component=self.component,
            environment=os.getenv("NODE_ENV") or os.getenv("ENVIRONMENT", "development"),
            service_name=os.getenv("K_SERVICE") or os.getenv("SERVICE_NAME"),
            service_version=os.getenv("K_REVISION") or os.getenv("SERVICE_VERSION"),
            custom=merged_context,
        )

        entry = LogEntry(
            level=level,
            message=message,
            timestamp=datetime.utcnow(),
            context=log_context,
            exception=exception,
            stack_trace=traceback.format_exc() if exception else None,
            source_file=frame.f_code.co_filename,
            source_line=frame.f_lineno,
            source_function=frame.f_code.co_name,
        )

        return entry

    def debug(self, message: str, **context) -> None:
        """Log a debug message."""
        entry = self._create_log_entry(LogLevel.DEBUG, message, **context)
        self.provider.log(entry)

    def info(self, message: str, **context) -> None:
        """Log an info message."""
        entry = self._create_log_entry(LogLevel.INFO, message, **context)
        self.provider.log(entry)

    def warning(self, message: str, **context) -> None:
        """Log a warning message."""
        entry = self._create_log_entry(LogLevel.WARNING, message, **context)
        self.provider.log(entry)

    def error(self, message: str, exception: Exception | None = None, **context) -> None:
        """Log an error message."""
        entry = self._create_log_entry(LogLevel.ERROR, message, exception, **context)
        self.provider.log(entry)

    def critical(self, message: str, exception: Exception | None = None, **context) -> None:
        """Log a critical message."""
        entry = self._create_log_entry(LogLevel.CRITICAL, message, exception, **context)
        self.provider.log(entry)

    def log_request(
        self,
        method: str,
        url: str,
        status: int | None = None,
        latency_ms: Optional[float] = None,
        response_size: int | None = None,
        user_agent: str | None = None,
        referer: str | None = None,
        **context,
    ) -> None:
        """Log HTTP request information."""
        level = (
            LogLevel.ERROR
            if status and status >= 500
            else LogLevel.WARNING if status and status >= 400 else LogLevel.INFO
        )

        message = f"{method} {url}"
        if status:
            message += f" -> {status}"

        entry = self._create_log_entry(level, message, **context)
        entry.http_method = method
        entry.http_url = url
        entry.http_status = status
        entry.http_latency_ms = latency_ms
        entry.http_response_size = response_size
        entry.http_user_agent = user_agent
        entry.http_referer = referer

        self.provider.log(entry)


class DefaultLoggerFactory(LoggerFactory):
    """
    Default logger factory implementation.

    Manages provider selection and creates logger instances.
    """

    def __init__(self, provider: LogProvider):
        self.provider = provider

    def create_logger(self, component: str, **default_context) -> Logger:
        """Create a logger for a specific component."""
        return DefaultLogger(component, self.provider, default_context)

    def set_provider(self, provider: LogProvider) -> None:
        """Set the logging provider."""
        if self.provider:
            self.provider.close()
        self.provider = provider

    def get_provider(self) -> LogProvider:
        """Get the current provider."""
        return self.provider


# Global factory instance
_logger_factory: Optional[LoggerFactory] = None


def create_logger_factory(provider: LogProvider) -> LoggerFactory:
    """
    Create a logger factory with the specified provider.

    This is typically called once during application startup.
    """
    global _logger_factory
    factory = DefaultLoggerFactory(provider)
    _logger_factory = factory
    return factory


def get_logger(component: str, **default_context) -> Logger:
    """
    Get a logger for the specified component.

    This is the main entry point for application code to get loggers.
    """
    if _logger_factory is None:
        raise RuntimeError("Logger factory not initialized. Call create_logger_factory() first.")

    return _logger_factory.create_logger(component, **default_context)


def set_provider(provider: LogProvider) -> None:
    """Set the global logging provider."""
    if _logger_factory is None:
        raise RuntimeError("Logger factory not initialized. Call create_logger_factory() first.")

    _logger_factory.set_provider(provider)
